src/types/dict.rs

```diff
@@ -557,76 +557,73 @@ pub(crate) use borrowed_iter::BorrowedDictIter;
 
 /// Conversion trait that allows a sequence of tuples to be converted into `PyDict`
 /// Primary use case for this trait is `call` and `call_method` methods as keywords argument.
-pub trait IntoPyDict: Sized {
+pub trait IntoPyDict<'py>: Sized {
     /// Converts self into a `PyDict` object pointer. Whether pointer owned or borrowed
     /// depends on implementation.
-    fn into_py_dict(self, py: Python<'_>) -> Bound<'_, PyDict>;
+    fn into_py_dict(self, py: Python<'py>) -> Bound<'_, PyDict>;
 
     /// Deprecated name for [`IntoPyDict::into_py_dict`].
     #[deprecated(since = "0.23.0", note = "renamed to `IntoPyDict::into_py_dict`")]
     #[inline]
-    fn into_py_dict_bound(self, py: Python<'_>) -> Bound<'_, PyDict> {
+    fn into_py_dict_bound(self, py: Python<'py>) -> Bound<'_, PyDict> {
         self.into_py_dict(py)
     }
 }
 
-impl<T, I> IntoPyDict for I
+impl<'py, T, I> IntoPyDict<'py> for I
 where
-    T: PyDictItem,
+    T: PyDictItem<'py>,
     I: IntoIterator<Item = T>,
 {
-    fn into_py_dict(self, py: Python<'_>) -> Bound<'_, PyDict> {
+    fn into_py_dict(self, py: Python<'py>) -> Bound<'_, PyDict> {
         let dict = PyDict::new(py);
         for item in self {
-            dict.set_item(item.key().to_object(py), item.value().to_object(py))
+            let (key, value) = item.unpack();
+            dict.set_item(key, value)
                 .expect("Failed to set_item on dict");
         }
         dict
     }
 }
 
 /// Represents a tuple which can be used as a PyDict item.
-pub trait PyDictItem {
-    type K: ToPyObject;
-    type V: ToPyObject;
-    fn key(&self) -> &Self::K;
-    fn value(&self) -> &Self::V;
+pub trait PyDictItem<'py> {
+    type K: IntoPyObject<'py>;
+    type V: IntoPyObject<'py>;
+    fn unpack(self) -> (Self::K, Self::V);
 }
```

Comment:
It's kind of breaking, but I can't really see a better way forward. Maybe we just mention this in the migration guide? Most users are not implementing / using this trait, I think.

Reply:
Actually I think `PyDictItem` is completely internal to PyO3 and is invisible from the outside (it's only mentioned inside `dict.rs` which is a `pub(crate)` module). So maybe we just remove `pub` here to make that clear? Also we should update the example that's already in the migration guide in that regard. I will fix it up together with the fallibility mention.

Reply:
I removed it and nothing complained, so it guess it's fine 😁